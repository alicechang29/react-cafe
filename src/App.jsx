import { useState, useEffect } from "react";
import { Route, Routes } from "react-router-dom";
import Home from "./Home";
import NavBar from "./NavBar";
import Menu from "./Menu";
import ItemLoader from "./ItemLoader";
import AddItemForm from "./AddItemForm";

import SnackOrBoozeApi from "./api";
import "./App.css";

/** Snack or Booze main application.
 *
 * Props: (none)
 *
 * State:
 * - itemsFetch:
 * {snacks: [item,item, ...], drinks: [item,item, ...], isLoading: true|false}
 *     where item is {id, name, description, recipe, serve}
 */

function App() {
  const [itemsFetch, setItemsFetch] = useState(
    {
      snacks: null,
      drinks: null,
      isLoading: true
    }
  );

  const { snacks, drinks, isLoading } = itemsFetch;
  console.log("* App %o", { snacks, drinks, isLoading });

  useEffect(function fetchItemsOnMount() {
    console.log("useEffect fetchItemsOnMount");

    fetchItems();
  }, []);

  if (isLoading) {
    return <p>Loading &hellip;</p>;
  }

  //fn to fetch items //TODO: docstring
  async function fetchItems() {
    const snacks = await SnackOrBoozeApi.getSnacks();
    const drinks = await SnackOrBoozeApi.getDrinks();
    setItemsFetch({ snacks, drinks, isLoading: false });
  }

  //fn to addItems to menu //TODO: docstring
  async function addItems(formData) {
    console.log("App - addItems", formData);

    await SnackOrBoozeApi.addItem(
      formData.menuItem,
      formData.type
    );

    //TODO: upon successful submission, navigate to...

    fetchItems();
  }

  return (
    <div className="App">
      <NavBar />
      <main>
        <Routes>
          <Route
            path="/"
            element={<Home snacks={snacks} drinks={drinks} />} />
          <Route
            path="/snacks"
            element={<Menu items={snacks} type="snacks" title="Snacks" />} />
          <Route
            path="/drinks"
            element={<Menu items={drinks} type="drinks" title="Drinks" />} />
          <Route
            path="/snacks/:id"
            element={<ItemLoader items={snacks} cantFindPath="/snacks" />} />
          <Route
            path="/drinks/:id"
            element={<ItemLoader items={drinks} cantFindPath="/drinks" />} />
          <Route
            path="/add-item"
            element={<AddItemForm addItem={addItems} />} />
          <Route
            path="*"
            element={<p>Hmmm. I can't seem to find what you want.</p>} />
        </Routes>
      </main>
    </div>
  );
}

export default App;
