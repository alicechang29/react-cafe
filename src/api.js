/** API calls for Snack or Booze.
 *
 * Static class tying together methods used to get/send to the API.
 */


class SnackOrBoozeAPI {
  static base_api_url = "http://localhost:5001";

  /** getSnacks
   * Fetch list of snacks
   * [{id, name, description, recipe, serve}, ...]
  */

  static async getSnacks() {
    console.log("getSnacks");
    const response = await fetch(`${this.base_api_url}/snacks`);
    return await response.json();
  }

  /** getDrinks
   * Fetch list of drinks
   * [{id, name, description, recipe, serve}, ...]
  */

  static async getDrinks() {
    console.log("getDrinks");
    const response = await fetch(`${this.base_api_url}/drinks`);
    return await response.json();
  }

  /** addItem
   * Adds item to the list of drinks or snacks
   * item: {id, name, description, recipe, serve}
   * type: "snacks" || "drinks"
  */

  static async addItem(item, type) {
    console.log("addItem");

    const response = await fetch(`${this.base_api_url}/${type}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ ...item })
    });
    return await response.json();
  }

}

export default SnackOrBoozeAPI;
