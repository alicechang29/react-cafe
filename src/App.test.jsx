import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";

import api from "./api";

import React from "react";
import { screen, fireEvent, render, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import App from "./App";

// api.getItems = vi.fn();
api.getDrinks = vi.fn();
api.getSnacks = vi.fn();
api.addItem = vi.fn();

const WELCOME_TEXT = "Welcome to Alice's Cafe!";

describe("Cafe routes", function () {
  beforeEach(function () {
    api.getDrinks.mockReturnValue([]);
    api.getSnacks.mockReturnValue([]);
    api.addItem.mockReturnValue([]);
  });

  afterEach(function () {
    api.getDrinks.mockReset([]);
    api.getSnacks.mockReset([]);
    api.addItem.mockReset();
  });

  it("renders without crashing", async function () {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => {
      screen.getByText(WELCOME_TEXT);
    });
  });

  it("renders the home snapshot", async function () {
    const { asFragment, debug } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText(WELCOME_TEXT));
    expect(api.getDrinks).toHaveBeenCalledTimes(1);
    expect(api.getSnacks).toHaveBeenCalledTimes(1);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a loading message before the data is loaded", async function () {
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    screen.getByText(/Loading\s*â€¦/i);
    await waitFor(() => screen.getByText(WELCOME_TEXT));
  });

  it("renders the snack menu", async function () {
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/snacks"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText(/Snacks\s*Menu/i));

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the drink menu", async function () {
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/drinks"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText(/Drinks\s*Menu/i));

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a drink from the menu", async function () {
    api.getDrinks.mockReturnValue([
      {
        id: "martini",
        name: "Martini",
        description: "A ice-cold, refreshing classic.",
        recipe: "Mix 3 parts vodka & 1 part dry vermouth.",
        serve: "Serve very cold, straight up.",
      },
    ]);
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/drinks/martini"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText("Martini"));
    screen.getByText(/refreshing classic/i);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a snack from the menu", async function () {
    api.getSnacks.mockReturnValue([
      {
        id: "nachos",
        name: "Nachos",
        description: "An American classic!",
        recipe: "Cover expensive, organic tortilla chips with Cheez Whiz.",
        serve:
          "Serve in a hand-thrown ceramic bowl, garnished with canned black olives",
      },
    ]);

    const { asFragment } = render(
      <MemoryRouter initialEntries={["/snacks/nachos"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText("Nachos"));
    screen.getByText(/an american classic/i);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the add item form", async function () {
    const { asFragment } = render(
      <MemoryRouter initialEntries={["/snacks/nachos"]}>
        <App />
      </MemoryRouter>
    );
    await waitFor(() => screen.getByText(/add item/i));
    expect(asFragment()).toMatchSnapshot();
  });
  //FIXME:
  // it("adds an item successfully", async function () {
  //   const { container, debug } = render(
  //     <MemoryRouter initialEntries={["/"]}>
  //       <App />
  //     </MemoryRouter>
  //   );
  //   await waitFor(() => fireEvent.click(screen.getByText("Add Item")));

  //   fireEvent.change(screen.getByLabelText("name"), {
  //     target: { value: "test name" },
  //   });
  //   fireEvent.change(screen.getByLabelText("description"), {
  //     target: { value: "test description" },
  //   });
  //   fireEvent.change(screen.getByLabelText("recipe"), {
  //     target: { value: "test recipe" },
  //   });
  //   fireEvent.change(screen.getByLabelText("serve"), {
  //     target: { value: "test serve" },
  //   });


  //   // // Submit the form
  //   // const button = screen.getByRole("button", { name: /Add to Menu!/i });
  //   // fireEvent.click(button);

  //   // // Verify the API call
  //   // await waitFor(() => expect(api.addItem).toHaveBeenCalled());

  //   // await waitFor(() => screen.getByText(WELCOME_TEXT));
  // });
});
