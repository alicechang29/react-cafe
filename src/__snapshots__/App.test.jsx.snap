// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cafe routes > renders a drink from the menu 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-md"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Alice's Cafe
        </a>
        <ul
          class="ms-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/snacks"
            >
              Snacks
            </a>
             
          </li>
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="active"
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
             
            <a
              class=""
              href="/add-item"
            >
              Add Item
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="RoutesList"
    >
      <main>
        <section
          class="Item"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <div
                class="fw-bold text-center card-title"
              >
                Martini
              </div>
              <p
                class="font-italic card-text"
              >
                A ice-cold, refreshing classic.
              </p>
              <p>
                <b>
                  Recipe:
                </b>
                 Mix 3 parts vodka & 1 part dry vermouth.
              </p>
              <p>
                <b>
                  Serve:
                </b>
                 Serve very cold, straight up.
              </p>
            </div>
          </div>
        </section>
      </main>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cafe routes > renders a snack from the menu 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-md"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Alice's Cafe
        </a>
        <ul
          class="ms-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="active"
              href="/snacks"
            >
              Snacks
            </a>
             
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
             
            <a
              class=""
              href="/add-item"
            >
              Add Item
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="RoutesList"
    >
      <main>
        <section
          class="Item"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <div
                class="fw-bold text-center card-title"
              >
                Nachos
              </div>
              <p
                class="font-italic card-text"
              >
                An American classic!
              </p>
              <p>
                <b>
                  Recipe:
                </b>
                 Cover expensive, organic tortilla chips with Cheez Whiz.
              </p>
              <p>
                <b>
                  Serve:
                </b>
                 Serve in a hand-thrown ceramic bowl, garnished with canned black olives
              </p>
            </div>
          </div>
        </section>
      </main>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cafe routes > renders the add item form 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-md"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Alice's Cafe
        </a>
        <ul
          class="ms-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="active"
              href="/snacks"
            >
              Snacks
            </a>
             
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
             
            <a
              class=""
              href="/add-item"
            >
              Add Item
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="RoutesList"
    >
      <main>
        <section
          class="col-md-4"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <div
                class="fw-bold text-center card-title"
              >
                SNACKS Menu
              </div>
              <p
                class="card-text"
              >
                Try one of our delicious menu items!
              </p>
              <ul
                class="list-group"
              />
            </div>
          </div>
        </section>
      </main>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cafe routes > renders the drink menu 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-md"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Alice's Cafe
        </a>
        <ul
          class="ms-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/snacks"
            >
              Snacks
            </a>
             
          </li>
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="active"
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
             
            <a
              class=""
              href="/add-item"
            >
              Add Item
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="RoutesList"
    >
      <main>
        <section
          class="col-md-4"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <div
                class="fw-bold text-center card-title"
              >
                DRINKS Menu
              </div>
              <p
                class="card-text"
              >
                Try one of our delicious menu items!
              </p>
              <ul
                class="list-group"
              />
            </div>
          </div>
        </section>
      </main>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cafe routes > renders the home snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-md"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Alice's Cafe
        </a>
        <ul
          class="ms-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/snacks"
            >
              Snacks
            </a>
             
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
             
            <a
              class=""
              href="/add-item"
            >
              Add Item
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="RoutesList"
    >
      <main>
        <section
          class="col-md-8"
        >
          <div
            class="card"
          >
            <div
              class="text-center card-body"
            >
              <div
                class="card-title"
              >
                <h3
                  class="fw-bold"
                >
                  Welcome to Alice's Cafe!
                </h3>
                <p>
                  Snacks: 0
                </p>
                <p>
                  Drinks: 0
                </p>
              </div>
            </div>
          </div>
        </section>
      </main>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cafe routes > renders the snack menu 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar navbar navbar-expand-md"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Alice's Cafe
        </a>
        <ul
          class="ms-auto navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="active"
              href="/snacks"
            >
              Snacks
            </a>
             
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/drinks"
            >
              Drinks
            </a>
          </li>
          <li
            class="nav-item"
          >
             
            <a
              class=""
              href="/add-item"
            >
              Add Item
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="RoutesList"
    >
      <main>
        <section
          class="col-md-4"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <div
                class="fw-bold text-center card-title"
              >
                SNACKS Menu
              </div>
              <p
                class="card-text"
              >
                Try one of our delicious menu items!
              </p>
              <ul
                class="list-group"
              />
            </div>
          </div>
        </section>
      </main>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches the screenshot 1`] = `
<div>
  <p>
    Loading …
  </p>
</div>
`;
